src/util.f90

```diff
@@ -152,100 +162,402 @@ SUBROUTINE MakeLiebMatrixStructrue(dm, nu, n, ucl, n_uc, nt, matr, cubesites, li
+
+     ! ------- For Lieb atoms except close to Cube atom of other unit cell-----------
+     IF(nu>1)THEN
+        DO k=1, nu-1
+
+           DO j=1, dm
+
+              ind = (i-1)*ucl + ucl_d(j) + k - 1
+              !print*,"Lieb site", ind
+              liebcount= liebcount+1
+              liebsites(liebcount)=ind
+
+              matr(ind,ind) = 0.0D0
+           END DO
+
+        END DO
+     END IF
+
+     ! For Lieb atoms close to Cube atoms of other unit cells
+
+     DO j=1,dm
+        IF(j==1)THEN
+           Flag = ( MOD(i,n**j)==0 )
+        ELSE IF(j==2)THEN
+           Flag = ( MOD(i,n**j)==0 .OR. (MOD(i,n**j).GT.(n**j-n)) )
+        ELSE
+           Flag = ( i.GT.(n**j-n**2) )
+        END IF
+
+        ind = (i-1)*ucl + ucl_d(j) + nu - 1
+        !print*,"Lieb site", ind
+        liebcount= liebcount+1
+        liebsites(liebcount)=ind
+
+        matr(ind, ind) = 0.0D0 
+     END DO
+
+  END DO
+!!!!! End construct Lieb Matrix
+
+  RETURN
+
+END SUBROUTINE MakeLiebOnsiteDisorderORIGINAL
+
+!-----------------------------------------------------------------
+SUBROUTINE MakeLiebOnsiteDisorder(dm, nu, n, ucl, n_uc, nt, matr, size, seed)
+
+  USE MyNumbers
+  USE IChannels
+  USE IPara
+  USE DPara
+
+  USE RNG_MT
+
+  IMPLICIT NONE
+
+  INTEGER(KIND=IKIND) &
+       dm, & ! the dimension
+       n, &  ! the number of unit cell in each dimension
+       nu, & ! the number of site in each edge
+       ucl, & ! the number of atoms in a unit cell
+       nt, &  ! the whole number of atoms in system
+       n_uc   ! the number of unit cells
+
+  INTEGER(KIND=IKIND) i, j, k, ind, cubecount,liebcount
+
+  LOGICAL(KIND=8) Flag
+
+  INTEGER(KIND=IKIND), ALLOCATABLE :: ucl_d(:) 
+  REAL(KIND=RKIND) matr(nt, nt)! , matr_W( nt, nt )
+  !INTEGER(KIND=IKIND) cubesites(n_uc), liebsites(nt-n_uc)
+
+  INTEGER seed(5), size, len,col,row 
+  REAL(KIND=RKIND) drandval, SUMCUBErandval,SUMLIEBrandval
+  !EXTERNAL DRANDOM5
+  
+  PRINT*,"MakeLiebOnsiteDisorder()"
+
+  !matr(:,:) = 0.0D0
+
+  ! ----------------------------------------------------------
+  ! Start construction for the DIAGONAL elements of the Lieb Matrix
+  ! ----------------------------------------------------------
+
+  SUMCUBErandval= 0.0D0; SUMLIEBrandval= 0.0D0
+
+  ! Give the Lieb matrix different onsite potentials
+  DO i=1, n_uc
+
+     ! CUBE onsite potentials
+     drandval= DRANDOM5(seed)
+
+     SUMCUBErandval=SUMCUBErandval + CubeDis*(drandval - 0.5D0)
+
+     SELECT CASE (IRNGFlag)
+     CASE(0) ! constant CubeConPot on each cube site
+        matr( (i-1)*ucl + 1 , (i-1)*ucl + 1 ) = CubeConPot + CubeDis*(drandval - 0.5D0)
+     CASE(1) ! +/- CubeConPot on random cube sites
+
+        IF(MOD(n_uc,2) .NE. 0) THEN
+           PRINT*, "main: WRNG, cube size is odd, so we cannot achieve 0 SUMMED potential!"
+           PRINT*, "main: WRNG, calculation will proceed, but output is questionable."
+        END IF
+
+        IF(DRANDOM5(seed)>=0.5) THEN
```

Comment:
The disorder does not fluctuate around CubeConPot but above +CubeConPot or below -CubeConpot, is it ok?

Reply:
I want the disorder to be CubeConPot * [ -CubeDis/2, +CubeDis/2 ] while CubeConPot can be constant for CPFlag=0, checkerboard +/- CubeConPot for CPFlag=1 and for CPFlag=2, the position of the checkerboard is random. I think this is what implemented here.